proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+Here we got a `Repository<User>` defined in a subpackage of `com.domain`, where the `User` type is defined.
+
+#### 5. Subpackages of the package where the constraint interface is defined
+
+Last place to look at would be subpackages of the package where the constraint interface is defined.
+
+```kotlin
+package com.data.instances
+
+import com.data.Repository
+import com.domain.User
+
+extension object UserRepository: Repository<User> {
+  override fun loadAll(): List<User> {
+    return listOf(User(25, "Bob"))
+  }
+
+  override fun loadById(id: Int): User? {
+    return if (id == 25) {
+      User(25, "Bob")
+    } else {
+      null
+    }
+  }
+}
+```
+
+Here, we are resolving it from `com.data.instances`, which a subpackage of `com.data`, where our constraint `Repository` is defined.
+
+## Error reporting
+
+We've got a `CallChecker` in place to report inlined errors using the context trace. That allows us to report as many errors as possible in a single compiler pass. Also provide them in two different formats:
+
+#### Inline errors while coding (using inspections and red underline)
+
+Whenever you're coding the checker is running and proper unresolvable extension errors can be reported within IDEA inspections.
+
+![Screenshot 2019-04-05 at 12 51 11](https://user-images.githubusercontent.com/6547526/55622809-85580480-57a1-11e9-8254-0830d29593ec.png)
+
+#### Errors once you hit the "compile" button:
+
+Once you hit the "compile" button or run any compile command you'll also get those errors reported.
+
+![Screenshot 2019-04-05 at 12 59 54](https://user-images.githubusercontent.com/6547526/55623259-be44a900-57a2-11e9-927e-fe0150265ba5.png)
+
+## How to try KEEEP-87? (First approach)
+
+- Clone [Our Kotlin fork](https://github.com/arrow-kt/kotlin) and checkout the **keep-87** branch.
+- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#build-environment-requirements) configure the necessary JVMs.
+- Follow the instructions on the [README](https://github.com/arrow-kt/kotlin/blob/master/ReadMe.md#-working-with-the-project-in-intellij-idea) to open the project in IntelliJ IDEA.
+- Once you have everything working, you can run a new instance of IntelliJ IDEA with the new modifications to the language by executing `./gradlew runIde`. There is also a pre-configured run configuration titled **IDEA** that does this.
+- It will open a new instance of the IDE where you can create a new project and experiment with the new features of the language. You can also download [this project](https://github.com/arrow-kt/kotlin/files/3064100/Keep87Sample.zip)
+) with some sample code that you can try out.
+
+## How to try KEEEP-87? (Alternative approach - easier)
```

Comment:
If this is the easier approach it should be listed first and the other one as the alternative.

Reply:
Done in 21aeaed.